runtime/src/iouring/mod.rs

```diff
@@ -160,6 +190,58 @@ pub(crate) async fn run(
     }
 }
 
+/// Process `ring` completions until all pending operations are complete or
+/// until `timeout` fires. If `timeout` is None, wait indefinitely.
+async fn drain(
+    ring: &mut IoUring,
+    waiters: &mut HashMap<u64, oneshot::Sender<i32>>,
+    has_op_timeout: bool,
+    timeout: Option<Duration>,
+) {
+    if let Some(timeout) = timeout {
+        // Create a timeout that will fire if we can't clear all the inflight operations.
+        let timeout = Timespec::new()
+            .sec(timeout.as_secs())
+            .nsec(timeout.subsec_nanos());
+        let timeout = Timeout::new(&timeout)
+            .build()
+            .user_data(SHUTDOWN_TIMEOUT_WORK_ID);
+        unsafe {
+            ring.submission()
+                .push(&timeout)
+                .expect("unable to push to queue");
+        }
+    }
+
+    while !waiters.is_empty() {
+        ring.submit_and_wait(1).expect("unable to submit to ring");
+        while let Some(cqe) = ring.completion().next() {
```

Comment:
This block is ~identical to the timeout handling in the loop. I wonder if there is a way to reuse this?